test/test_type_promotion.py

```diff
@@ -218,6 +218,47 @@ def test_create_bool_tensors(self):
         # this seems like odd behavior but ints also create float tensors, numpy doesn't have this function.
         self.assertEqual(torch.scalar_tensor(False, device=self.device), torch.tensor(0., device=self.device))
 
+    def test_comparison_ops_with_type_promotion(self):
+        value_for_type = {
+            torch.uint8: (1 << 5),
+            torch.int8: (1 << 5),
+            torch.int16: (1 << 10),
+            torch.int32: (1 << 20),
+            torch.int64: (1 << 30),
+            torch.float16: (1 << 10),
+            torch.float32: (1 << 20),
+            torch.float64: (1 << 30)
+        }
+        comparison_ops = [
+            dict(
+                name="lt",
+                out_op=lambda x, y, d: torch.lt(x, y, out=torch.empty(1, dtype=torch.bool, device=d)),
+                ret_op=lambda x, y: torch.lt(x, y),
+                compare_op=lambda x, y: x < y,
+            ),
+        ]
+        for device in torch.testing.get_all_device_types():
```

Comment:
nit: Use `itertools.product` to avoid the indentation:

```
for device, op, dt1, dt2 in itertools.product(torch.testing.get_all_device_types(), comparison_ops, 
                                          torch.testing.get_all_math_dtypes(device), 
                                          torch.testing.get_all_math_dtypes(device)
```

Reply:
I think in this case it's impossible because of get_all_math_dtypes(device): local variable 'device' referenced before assignment

Reply:
Well, you can itertools.product the last three loops instead. But again this is a nit